cpp/ql/src/Security/CWE/CWE-014/MemsetMayBeDeleted.ql

```diff
@@ -0,0 +1,50 @@
+/**
+ * @name Call to `memset` may be deleted
+ * @description Using <code>memset</code> the function to clear private data in a variable that has no subsequent use
+ *              is potentially dangerous because the compiler can remove the call.
+ * @kind problem
+ * @id cpp/memset-may-be-deleted
+ * @problem.severity warning
+ * @precision high
+ * @tags security
+ *       external/cwe/cwe-14
+ */
+
+import cpp
+import semmle.code.cpp.dataflow.EscapesTree
+import semmle.code.cpp.commons.Exclusions
+
+class MemsetFunction extends Function {
+  MemsetFunction() {
+    this.hasGlobalOrStdOrBslName("memset")
+    or
+    this.hasGlobalOrStdName("wmemset")
+    or
+    this.hasGlobalName(["bzero", "__builtin_memset"])
+  }
+}
+
+from FunctionCall call, LocalVariable v, MemsetFunction memset
+where
+  call.getTarget() = memset and
+  not isFromMacroDefinition(call) and
+  // `v` only escapes as the argument to `memset`.
+  forall(Expr escape | variableAddressEscapesTree(v.getAnAccess(), escape) |
+    call.getArgument(0) = escape.getUnconverted()
+  ) and
+  // `v` is a stack-allocated array or a struct, and `v` is not static.
+  not v.isStatic() and
+  (
+    v.getUnspecifiedType() instanceof ArrayType and call.getArgument(0) = v.getAnAccess()
+    or
+    v.getUnspecifiedType() instanceof Struct and
+    call.getArgument(0).(AddressOfExpr).getAddressable() = v
+  ) and
```

Comment:
I think this is a syntactic way of writing `variableAddressEscapesTree(v.getAnAccess(), call.getArgument(0).getFullyConverted())`, but there might be differences in edge cases I'm not grasping.

Reply:
Indeed, `variableAddressEscapesTree` is really what I wanted. But still with the restriction that the type is an array or struct, I think.

Reply:
What do we get from the restriction to array or struct?

Reply:
I was afraid of getting false positives in cases such as:
```cpp
struct S2;

struct S1 { S2* ps2; };

void test_abc(S1* ps1) {
  S2* ps2 = ps1->ps2;
  // ...
  memset(ps2, 0, sizeof(S2));
}
```

but it seems like `variableAddressEscapesTree` also rules out these. So I guess that restriction isn't actually needed then 🎉!

Reply:
We should probably rule out reference-typed variables since `variableAddressEscapesTree` has special behaviour on those.